File: pulsar-core/pulsar-tools/pulsar-browser/src/main/java/com/github/kklisura/cdt/protocol/v2023/commands/Tracing.java
package com.github.kklisura.cdt.protocol.v2023.commands;

/*-
 * #%L
 * cdt-java-client
 * %%
 * Copyright (C) 2018 - 2023 Kenan Klisura
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.github.kklisura.cdt.protocol.v2023.events.tracing.BufferUsage;
import com.github.kklisura.cdt.protocol.v2023.events.tracing.DataCollected;
import com.github.kklisura.cdt.protocol.v2023.events.tracing.TracingComplete;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.*;
import com.github.kklisura.cdt.protocol.v2023.support.types.EventHandler;
import com.github.kklisura.cdt.protocol.v2023.support.types.EventListener;
import com.github.kklisura.cdt.protocol.v2023.types.tracing.*;

import java.util.List;

@Experimental
public interface Tracing {

  /** Stop trace events collection. */
  void end();

  /** Gets supported tracing categories. */
  @Returns("categories")
  @ReturnTypeParameter(String.class)
  List<String> getCategories();

  /**
   * Record a clock sync marker in the trace.
   *
   * @param syncId The ID of this clock sync marker
   */
  void recordClockSyncMarker(@ParamName("syncId") String syncId);

  /** Request a global memory dump. */
  RequestMemoryDump requestMemoryDump();

  /**
   * Request a global memory dump.
   *
   * @param deterministic Enables more deterministic results by forcing garbage collection
   * @param levelOfDetail Specifies level of details in memory dump. Defaults to "detailed".
   */
  RequestMemoryDump requestMemoryDump(
      @Optional @ParamName("deterministic") Boolean deterministic,
      @Optional @ParamName("levelOfDetail") MemoryDumpLevelOfDetail levelOfDetail);

  /** Start trace events collection. */
  void start();

  /**
   * Start trace events collection.
   *
   * @param categories Category/tag filter
   * @param options Tracing options
   * @param bufferUsageReportingInterval If set, the agent will issue bufferUsage events at this
   *     interval, specified in milliseconds
   * @param transferMode Whether to report trace events as series of dataCollected events or to save
   *     trace to a stream (defaults to `ReportEvents`).
   * @param streamFormat Trace data format to use. This only applies when using `ReturnAsStream`
   *     transfer mode (defaults to `json`).
   * @param streamCompression Compression format to use. This only applies when using
   *     `ReturnAsStream` transfer mode (defaults to `none`)
   * @param traceConfig
   * @param perfettoConfig Base64-encoded serialized perfetto.protos.TraceConfig protobuf message
   *     When specified, the parameters `categories`, `options`, `traceConfig` are ignored. (Encoded
   *     as a base64 string when passed over JSON)
   * @param tracingBackend Backend type (defaults to `auto`)
   */
  void start(
      @Deprecated @Optional @ParamName("categories") String categories,
      @Deprecated @Optional @ParamName("options") String options,
      @Optional @ParamName("bufferUsageReportingInterval") Double bufferUsageReportingInterval,
      @Optional @ParamName("transferMode") StartTransferMode transferMode,
      @Optional @ParamName("streamFormat") StreamFormat streamFormat,
      @Optional @ParamName("streamCompression") StreamCompression streamCompression,
      @Optional @ParamName("traceConfig") TraceConfig traceConfig,
      @Optional @ParamName("perfettoConfig") String perfettoConfig,
      @Optional @ParamName("tracingBackend") TracingBackend tracingBackend);

  @EventName("bufferUsage")
  EventListener onBufferUsage(EventHandler<BufferUsage> eventListener);

  /**
   * Contains a bucket of collected trace events. When tracing is stopped collected events will be
   * sent as a sequence of dataCollected events followed by tracingComplete event.
   */
  @EventName("dataCollected")
  EventListener onDataCollected(EventHandler<DataCollected> eventListener);

  /**
   * Signals that tracing is stopped and there is no trace buffers pending flush, all data were
   * delivered via dataCollected events.
   */
  @EventName("tracingComplete")
  EventListener onTracingComplete(EventHandler<TracingComplete> eventListener);
}

File: pulsar-core/pulsar-tools/pulsar-browser/src/main/kotlin/ai/platon/cdt/kt/protocol/commands/Tracing.kt
@file:Suppress("unused")
package ai.platon.cdt.kt.protocol.commands

import ai.platon.cdt.kt.protocol.events.tracing.BufferUsage
import ai.platon.cdt.kt.protocol.events.tracing.DataCollected
import ai.platon.cdt.kt.protocol.events.tracing.TracingComplete
import ai.platon.cdt.kt.protocol.support.annotations.EventName
import ai.platon.cdt.kt.protocol.support.annotations.Experimental
import ai.platon.cdt.kt.protocol.support.annotations.Optional
import ai.platon.cdt.kt.protocol.support.annotations.ParamName
import ai.platon.cdt.kt.protocol.support.annotations.ReturnTypeParameter
import ai.platon.cdt.kt.protocol.support.annotations.Returns
import ai.platon.cdt.kt.protocol.support.types.EventHandler
import ai.platon.cdt.kt.protocol.support.types.EventListener
import ai.platon.cdt.kt.protocol.types.tracing.MemoryDumpLevelOfDetail
import ai.platon.cdt.kt.protocol.types.tracing.RequestMemoryDump
import ai.platon.cdt.kt.protocol.types.tracing.StartTransferMode
import ai.platon.cdt.kt.protocol.types.tracing.StreamCompression
import ai.platon.cdt.kt.protocol.types.tracing.StreamFormat
import ai.platon.cdt.kt.protocol.types.tracing.TraceConfig
import ai.platon.cdt.kt.protocol.types.tracing.TracingBackend
import kotlin.Boolean
import kotlin.Double
import kotlin.String
import kotlin.Unit
import kotlin.collections.List

@Experimental
interface Tracing {
  /**
   * Stop trace events collection.
   */
  suspend fun end()

  /**
   * Gets supported tracing categories.
   */
  @Returns("categories")
  @ReturnTypeParameter(String::class)
  suspend fun getCategories(): List<String>

  /**
   * Record a clock sync marker in the trace.
   * @param syncId The ID of this clock sync marker
   */
  suspend fun recordClockSyncMarker(@ParamName("syncId") syncId: String)

  /**
   * Request a global memory dump.
   * @param deterministic Enables more deterministic results by forcing garbage collection
   * @param levelOfDetail Specifies level of details in memory dump. Defaults to "detailed".
   */
  suspend fun requestMemoryDump(@ParamName("deterministic") @Optional deterministic: Boolean? = null, @ParamName("levelOfDetail") @Optional levelOfDetail: MemoryDumpLevelOfDetail? = null): RequestMemoryDump

  suspend fun requestMemoryDump(): RequestMemoryDump {
    return requestMemoryDump(null, null)
  }

  /**
   * Start trace events collection.
   * @param categories Category/tag filter
   * @param options Tracing options
   * @param bufferUsageReportingInterval If set, the agent will issue bufferUsage events at this interval, specified in milliseconds
   * @param transferMode Whether to report trace events as series of dataCollected events or to save trace to a
   * stream (defaults to `ReportEvents`).
   * @param streamFormat Trace data format to use. This only applies when using `ReturnAsStream`
   * transfer mode (defaults to `json`).
   * @param streamCompression Compression format to use. This only applies when using `ReturnAsStream`
   * transfer mode (defaults to `none`)
   * @param traceConfig
   * @param perfettoConfig Base64-encoded serialized perfetto.protos.TraceConfig protobuf message
   * When specified, the parameters `categories`, `options`, `traceConfig`
   * are ignored. (Encoded as a base64 string when passed over JSON)
   * @param tracingBackend Backend type (defaults to `auto`)
   */
  suspend fun start(
    @ParamName("categories") @Optional categories: String? = null,
    @ParamName("options") @Optional options: String? = null,
    @ParamName("bufferUsageReportingInterval") @Optional bufferUsageReportingInterval: Double? = null,
    @ParamName("transferMode") @Optional transferMode: StartTransferMode? = null,
    @ParamName("streamFormat") @Optional streamFormat: StreamFormat? = null,
    @ParamName("streamCompression") @Optional streamCompression: StreamCompression? = null,
    @ParamName("traceConfig") @Optional traceConfig: TraceConfig? = null,
    @ParamName("perfettoConfig") @Optional perfettoConfig: String? = null,
    @ParamName("tracingBackend") @Optional tracingBackend: TracingBackend? = null,
  )

  suspend fun start() {
    return start(null, null, null, null, null, null, null, null, null)
  }

  @EventName("bufferUsage")
  fun onBufferUsage(eventListener: EventHandler<BufferUsage>): EventListener

  @EventName("bufferUsage")
  fun onBufferUsage(eventListener: suspend (BufferUsage) -> Unit): EventListener

  @EventName("dataCollected")
  fun onDataCollected(eventListener: EventHandler<DataCollected>): EventListener

  @EventName("dataCollected")
  fun onDataCollected(eventListener: suspend (DataCollected) -> Unit): EventListener

  @EventName("tracingComplete")
  fun onTracingComplete(eventListener: EventHandler<TracingComplete>): EventListener

  @EventName("tracingComplete")
  fun onTracingComplete(eventListener: suspend (TracingComplete) -> Unit): EventListener
}
